### PR TITLE
MLIR#39 - Adopt the new amdgpu buffer raw intrinsics (phase 1)

### PR DESCRIPTION
**Pull request by sjw36**
- Added the intrinsics to ROCDLOps
- Added translation to llvm.amdgcn
- Added tests

### CODE REVIEW

**Comment by ROCm-Apps-Test** [comment]
Can one of the admins verify this patch?

**Comment by sjw36** [comment]
see https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/39

File: mlir/test/Target/rocdl.mlir [hunk 1]
```diff
@@ -177,6 +177,28 @@ llvm.func @rocdl.mubuf(%rsrc : !llvm<"<4 x i32>">, %vindex : !llvm.i32,
...
+  %r2 = rocdl.raw.buffer.load %rsrc, %offset, %soffset, %aux : !llvm<"<2 x float>">
+  // CHECK: call <4 x float> @llvm.amdgcn.raw.buffer.load.v4f32(<4 x i32> %{{.*}}, i32 %{{.*}}, i32 %{{.*}}, i32 %{{.*}})
+  %r4 = rocdl.raw.buffer.load %rsrc, %offset, %soffset, %aux : !llvm<"<4 x float>">
+
+  // CHECK: call void @llvm.amdgcn.raw.buffer.store.v1f32(<1 x float> %{{.*}}, <4 x i32> %{{.*}}, i32 %{{.*}}, i32 %{{.*}}, i32 %{{.*}})
+  rocdl.raw.buffer.store %vdata1, %rsrc, %offset, %soffset, %aux : !llvm<"<1 x float>">
+  // CHECK: call void @llvm.amdgcn.raw.buffer.store.v2f32(<2 x float> %{{.*}}, <4 x i32> %{{.*}}, i32 %{{.*}}, i32 %{{.*}}, i32 %{{.*}})
+  rocdl.raw.buffer.store %vdata2, %rsrc, %offset, %soffset, %aux : !llvm<"<2 x float>">
+  // CHECK: call void @llvm.amdgcn.raw.buffer.store.v4f32(<4 x float> %{{.*}}, <4 x i32> %{{.*}}, i32 %{{.*}}, i32 %{{.*}}, i32 %{{.*}})
+  rocdl.raw.buffer.store %vdata4, %rsrc, %offset, %soffset, %aux : !llvm<"<4 x float>">
```
Comment (jerryyin) [inline mlir/test/Target/rocdl.mlir:197]: @asroy: Wants to make sure those exposed intrinsics are what you guys needed?

**Review by jerryyin — APPROVED** [review]
Adding a pointers to: [AMDGPURawBufferLoad](https://github.com/whchung/llvm-project/blob/master/llvm/include/llvm/IR/IntrinsicsAMDGPU.td#L890-L909), [AMDGPURawBufferStore](https://github.com/whchung/llvm-project/blob/master/llvm/include/llvm/IR/IntrinsicsAMDGPU.td#L926-L939).

I've got to say this raw buffer load intrinsic looks almost exactly the same to `MubufLoad`.

**Comment by sjw36** [comment]
jenkins : retest this please

**Comment by asroy** [comment]
@sjw36 How are these raw buffer intrinsics tested? I've never tested them in HIP so I don't know if compiler backend worked correctly with them or not

**Comment by sjw36** [comment]
@asroy, the tests I added only generate the llvm IR. I presumed the backend was tested already.. I will add one for that also.

File: mlir/test/Target/rocdl.mlir (hunk 1 above)
Comment (asroy) [reply mlir/test/Target/rocdl.mlir:197]: Yes. In the future, raw buffer atomic add for fp32 will also be needed.

Comment (jerryyin) [reply mlir/test/Target/rocdl.mlir:197]: @sjw36 Could you help create a new issue in `llvm-project-private` to track the atomic add? Reference this review and @asroy in the new issue.

Comment (jerryyin) [reply mlir/test/Target/rocdl.mlir:197]: Issue submitted [here](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/61).

**Pull request merged**